core/pta/device.c

```diff
@@ -38,46 +48,24 @@ static TEE_Result get_devices(uint32_t types,
 	if (!params[0].memref.buffer && (params[0].memref.size > 0))
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	device_uuid = (TEE_UUID *)params[0].memref.buffer;
-	ip_size = params[0].memref.size;
-
-	SCATTERED_ARRAY_FOREACH(ta, pseudo_tas, struct pseudo_ta_head) {
-		if (ta->flags & TA_FLAG_DEVICE_ENUM) {
-			if (ip_size < sizeof(TEE_UUID)) {
-				res = TEE_ERROR_SHORT_BUFFER;
-			} else {
-				tee_uuid_to_octets(uuid_octet, &ta->uuid);
-				memcpy(device_uuid, uuid_octet,
-				       sizeof(TEE_UUID));
-				device_uuid++;
-				ip_size -= sizeof(TEE_UUID);
-			}
-			op_size += sizeof(TEE_UUID);
-		}
-	}
+	buf =  params[0].memref.buffer;
+	blen = params[0].memref.size;
+
+	SCATTERED_ARRAY_FOREACH(ta, pseudo_tas, struct pseudo_ta_head)
+		add_ta(ta->flags, &ta->uuid, buf, blen, &pos);
 
 	if (IS_ENABLED(CFG_EARLY_TA)) {
 		const struct early_ta *eta = NULL;
```

Comment:
minor suggestion:
```c
static TEE_Result get_devices(uint32_t types,
			      TEE_Param params[TEE_NUM_PARAMS])
{
	const struct pseudo_ta_head *pta = NULL;
	const struct early_ta *eta = NULL;
	(...)

	SCATTERED_ARRAY_FOREACH(pta, pseudo_tas, struct pseudo_ta_head)
		add_ta(pta->flags, &pta->uuid, buf, blen, &pos);

	if (IS_ENABLED(CFG_EARLY_TA))
		for_each_early_ta(eta)
			add_ta(eta->flags, &eta->uuid, buf, blen, &pos);

	(...)
}
```

Reply:
> There are still too many problems with this PR. I took the liberty to fix them myself and open a new PR: #3782. Please help review.
> 
> @muvarov could you check if everything still works for you? Thanks.

@jforissier  Big thanks that is quite annoying to correct patches descriptions and reorder them. These comments can never end. I will try to test tomorrow morning that patches from #3782 still work.

Reply:
@muvarov you're welcome. Indeed the review process is tedious sometimes, but you get used to it eventually :wink: and it is essential to code quality so we can be thankful to reviewers for their time and to contributors who bare with them :smile: